packages/database/src/changeHandlers/ChangeHandler.js

```diff
@@ -0,0 +1,115 @@
+   *
+   * If no pusher is defined for the record type, then the `changeDetails` generated by the db
+   * will be simply added to the change queue
+   *
+   * @protected
+   * @type {Record<string, Function>}
+   */
+  changePushers = {};
+
+  /**
+   * A map of change handlers by record type. Each handler will be used to process queued changes
+   * of that type
+   *
+   * @protected
+   * @type {Record<string, Function>}
+   */
+  changeHandlers = {};
+
+  /**
+   * Wait 1 sec after changes before handling the change queue, to avoid double-up.
+   * Can override in child classes to fine-tune the timing of queue handling
+   *
+   * @protected
+   */
+  debounceTime = 1000; // ms
+
+  /**
+   * @type {Record<string, Array>} Changes by record type
+   */
+  changeQueue = {};
+
+  scheduledTimeout = null;
+
+  scheduledPromise = null;
+
+  scheduledPromiseResolve = null;
+
+  activePromise = null;
+
+  changeHandlerCancellers = [];
+
+  constructor(models) {
+    this.models = models;
+  }
+
+  setDebounceTime(debounceTime) {
+    this.debounceTime = debounceTime;
+  }
+
+  listenForChanges() {
+    this.changeHandlerCancellers = Object.keys(this.changeHandlers).map(recordType => {
+      this.changeQueue[recordType] = [];
+      const changePusher = this.changePushers[recordType] || (changeDetails => [changeDetails]);
+
+      const pushAndSchedule = async changeDetails => {
+        this.changeQueue[recordType].push(...changePusher(changeDetails));
+        // Changes are scheduled to be handled as a batch at a later stage
+        return this.scheduleChangeQueueHandler(recordType);
+      };
+      return this.models[recordType].addChangeHandler(pushAndSchedule);
+    });
+  }
+
+  stopListeningForChanges() {
+    this.changeHandlerCancellers.forEach(c => c());
+    this.changeHandlerCancellers = [];
+  }
+
+  async scheduleChangeQueueHandler(recordType) {
+    // wait for any active refresh to finish before scheduling a new one
+    await this.activePromise;
+
+    // clear any previous scheduled handler, so that we debounce all changes in the same time period
+    if (this.scheduledTimeout) {
+      clearTimeout(this.scheduledTimeout);
+    }
+
+    if (!this.scheduledPromise) {
+      this.scheduledPromise = new Promise(resolve => {
+        this.scheduledPromiseResolve = resolve;
+      });
+    }
+
+    // schedule the handler to execute after an adequate period of debouncing
+    this.scheduledTimeout = setTimeout(() => {
+      this.activePromise = this.handleChangeQueue(recordType);
+    }, this.debounceTime);
+
+    return this.scheduledPromise;
+  }
+
+  handleChangeQueue = async recordType => {
+    // remove timeout so any changes added now get scheduled anew
+    this.scheduledTimeout = null;
+    this.scheduledPromise = null;
+
+    const changeQueueHandler = this.changeHandlers[recordType];
+    assert.ok(changeQueueHandler);
+
+    await changeQueueHandler(this.changeQueue[recordType]);
+    this.changeQueue[recordType] = [];
```

Comment:
I believe this could be a race condition, where the listeners could add new records to the change queue while the existing batch of changes is being processed. Those new ones would then be cleared off the change queue when the existing batch finishes processing, despite not having been processed themselves.

Fix could look something like:
```suggestion
    const queuedChanges = this.changeQueue[recordType];
    this.changeQueue[recordType] = [];
    try {
      await changeQueueHandler(queuedChanges);
    } catch (e) {
      this.changeQueue.push(...queuedChanges);
      throw e;
    }
```

Reply:
Awesome observation @edmofro !! You are really the async queue processing master ⏳🐱‍👤💯

Reply:
One potential issue here @edmofro is that if a batch of changes fails for a fundamental reason (eg error in code), we'll keep on pushing the same changes in the queue and the handling will fail again and again

I guess a solution would be implementing a more elaborate queue management system (similar to `dhis_sync_queue`)... It does sounds a bit overengineered for what we want to achieve now, but ignoring those changes completely is not ideal either.

Reply:
I'm thinking that this is an existing issue of handling change notifications: if the information is not persisted in the DB, we'll have no way of knowing that a change handling failed (and needs to run again) apart from checking the logs.

One solution would be adding a `change_handler_success` column in `survey_response`, which is set to `false` if something goes wrong in a change handler. We can then ignore the records whose change handling failed, and try to re-handle them in the future after the issue is fixed. This is a simplistic approach which assumes that only one change handler is defined for `survey_response`. It requires refactoring the change translators to return a map of db records to change queue items.

I'm not a huge fan of the above, but may be a simple compromise between simplicity and accuracy. Of course, the other option is ignoring this potential danger for now, and deal with bugs if and when they arise 🔥

Reply:
I'd go for ignoring the potential danger, and letting the change in the queue fail again and again. The mitigation for that would be something we need to do anyway: logging to an external service that will do things like group logs and allow us to explore them easily, and potentially notify us in unexpected situations. I believe @IgorNadj has started some work looking at external logging services?

Reply:
Yep we have https://github.com/beyondessential/tupaia-backlog/issues/2003. I never got to it this sprint. Still something we should do but will take a good few days to set up properly.

Reply:
We were discussing about this with @IgorNadj and he suggested the following: for each failed batch, re-try a max number of times (say 3), if the batch fails then log some info and stop retrying

With that in mind, I suggest using a `logChangeInfo` to log debugging info when we keep re-trying for a failed batch. By default it would just log the count of changes in a batch. It can be overridden in child classes to extract more useful debugging info. For example, `SurveyResponseOutdater` could extract the record ids of survey responses in a failed batch.

Thoughts @edmofro ?

Reply:
It sounds like a good pattern to me. Obviously it would also be great to get you onto viz builder as soon as we can, as it's just Chris once Tom moves back to NZ after tomorrow, so please don't spend too long on it, but what you suggest sounds like it shouldn't take too long :-)

Reply:
It is what it is - I agree with the priorities we have set and ideally we wouldn't want to spend too much time on this. On the other hand, we've been burned by invalid DB state creeping without us being aware (eg DHIS responses that were not added in the sync queue), so I think we should be careful given this will affect the core of our system